codalab/rest/bundle.py

```diff
@@ -1,19 +1,517 @@
+
+        # Save bundle into model
+        local.model.save_bundle_rest(bundle)
+
+        # Inherit worksheet permissions
+        group_permissions = local.model.get_group_worksheet_permissions(
+            request.user.user_id, worksheet_uuid)
+        set_bundle_permissions_helper([{
+            'object_uuid': bundle_uuid,
+            'group_uuid': p['group_uuid'],
+            'permission': p['permission'],
+        } for p in group_permissions])
+
+        # Add as item to worksheet
+        if shadow_parent_uuid is None:
+            local.model.add_worksheet_item(
+                worksheet_uuid, worksheet_util.bundle_item(bundle_uuid))
+        else:
+            local.model.add_shadow_worksheet_items(
+                shadow_parent_uuid, bundle_uuid)
+
+    # Get updated bundles
+    bundles_dict = get_bundle_infos(created_uuids)
+
+    # Return list of bundles in original order
+    return [bundles_dict[uuid] for uuid in created_uuids]
+
+
+@patch('/bundles', apply=AuthenticatedPlugin())
+def update_bundles():
+    bundle_updates = BundleSchema(
+        strict=True,
+        many=True,
+        dump_only=UPDATE_RESTRICTED_FIELDS,
+    ).load(request.json, partial=True).data
+    bundles = update_bundles_helper(bundle_updates)
+    return BundleSchema(many=True).dump(bundles).data
+
+
+def update_bundles_helper(bundle_updates):
+    """
+    Update bundle owners and/or metadata.
+    """
+    # Check permissions
+    bundle_uuids = [b['uuid'] for b in bundle_updates]
+    check_bundles_have_all_permission(local.model, request.user, bundle_uuids)
+
+    # Check that bundle_types match original
+    bundles_dict = get_bundle_infos(bundle_uuids)
+    for update in bundle_updates:
+        if ('bundle_type' in update and
+                    update['bundle_type'] != bundles_dict[update['uuid']]['bundle_type']):
+            abort(httplib.FORBIDDEN, "Updating bundle_type is forbidden")
+
+    # Update bundles
+    for update in bundle_updates:
+        # Prep bundle and save to model
+        for dep in update.get('metadata', []):
+            dep['bundle_uuid'] = update['uuid']
+        local.model.update_bundle_rest(update)
+
+    # Get updated bundles
+    bundles_dict = get_bundle_infos(bundle_uuids)
+
+    # Create list of bundles in original order
+    bundles = [bundles_dict[uuid] for uuid in bundle_uuids]
+
+    return bundles
+
+
+@delete('/bundles', apply=AuthenticatedPlugin())
+def delete_bundles():
+    """
+    Delete the bundles specified.
+    If |force|, allow deletion of bundles that have descendants or that appear across multiple worksheets.
+    If |recursive|, add all bundles downstream too.
+    If |data-only|, only remove from the bundle store, not the bundle metadata.
+    If |dry-run|, just return list of bundles that would be deleted, but do not actually delete.
+    """
+    uuids = get_resource_ids(request.json, 'bundles')
+    force = query_get_bool('force', default=False)
```

Comment:
When/why did we remove the ability to force delete bundles that are
dependencies of other bundles?

On Thu, May 26, 2016 at 11:05 AM, Konstantin Lopyrev <
notifications@github.com> wrote:

> In codalab/rest/bundle.py
> <https://github.com/codalab/codalab-cli/pull/433#discussion_r64792304>:
>
> > +    bundles = [bundles_dict[uuid] for uuid in bundle_uuids]
> > +
> > +    return bundles
> > +
> > +
> > +@delete('/bundles', apply=AuthenticatedPlugin())
> > +def delete_bundles():
> > +    """
> > +    Delete the bundles specified.
> > +    If |force|, allow deletion of bundles that have descendants or that appear across multiple worksheets.
> > +    If |recursive|, add all bundles downstream too.
> > +    If |data-only|, only remove from the bundle store, not the bundle metadata.
> > +    If |dry-run|, just return list of bundles that would be deleted, but do not actually delete.
> > +    """
> > +    uuids = get_resource_ids(request.json, 'bundles')
> > +    force = query_get_bool('force', default=False)
>
> Force is deprecated. Please look at the new logic in the delete_bundles
> method, specifically the code under self.launch_new_worker_system.
>
> You don't need to implement the if not force: path. The force flag will be
> removed.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/433/files/e4cc91a21049587251778b2aebe1e2a7cfeb5428#r64792304>
>